### PR TITLE
feat(aws-cognito): send emails with a verified domain

### PR DESCRIPTION
**Pull request by hassanazharkhan**
When sending emails with a verified domain, the email address does not need to be verified.
In that case, the identity of the SourceArn in EmailConfiguration is allowed to be set to the domain instead of the email address.

closes [#19762](https://github.com/aws/aws-cdk/issues/19762)

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `cdk-integ` to deploy the infrastructure and generate the snapshot (i.e. `cdk-integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/test/user-pool.test.ts [hunk 1]
```diff
@@ -1729,6 +1729,51 @@ describe('User Pool', () => {
...
+              {
+                Ref: 'AWS::Partition',
+              },
+              ':ses:us-east-1:11111111111:identity/example.com',
+            ],
+          ],
+        },
+      },
+    });
+
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:1774]: Can we remove this blank line?

File: packages/@aws-cdk/aws-cognito/test/user-pool.test.ts [hunk 2]
```diff
@@ -1729,6 +1729,51 @@ describe('User Pool', () => {
...
+  });
+
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:1776]: Can we remove this blank line?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts [hunk 3]
```diff
@@ -56,6 +56,13 @@ export interface UserPoolSESOptions {
    * @default - The same region as the Cognito UserPool
    */
   readonly sesRegion?: string;
+
+  /**
+   * SES Verified custom domain to be used to verify the identity
+   *
+   * @default - no domain
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts:63]: should be

```
@default no domain
```

(I know the others in this file are written with the `-`, but they shouldn't be. You can remove the `-` from the others here as well if you want to).

**Review by comcalvi — CHANGES_REQUESTED** [review]
Looks great! A few minor comments.

**Comment by corymhall** [comment]
>> New Features
Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
Did you use cdk-integ to deploy the infrastructure and generate the snapshot (i.e. cdk-integ without --dry-run)?

@hassanazharkhan this would be considered a new feature. Can you add a new integration test for this?

**Comment by hassanazharkhan** [comment]
> > > New Features
> > > Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
> > > Did you use cdk-integ to deploy the infrastructure and generate the snapshot (i.e. cdk-integ without --dry-run)?
> 
> @hassanazharkhan this would be considered a new feature. Can you add a new integration test for this?

I was considering this as fix as with current implementation, I'm not able to use SES domain identity, since you're considering this a feat, then It'd definitely require inter.test and README changes.

> > > New Features
> > > Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
> > > Did you use cdk-integ to deploy the infrastructure and generate the snapshot (i.e. cdk-integ without --dry-run)?
> 
> @hassanazharkhan this would be considered a new feature. Can you add a new integration test for this?

Sure, let me add it

**Comment by kt-hr** [comment]
Hello, @hassanazharkhan.
Thank you for fixing this. Is there any update?
I have the same modification in #19878 and hope it helps.

**Comment by hassanazharkhan** [comment]
@kt-hr Tye only thing I need to add is the integ.test and after that this one will be good to merge, I'll try to do that today.

At the same time I'm happy to close this one if you'd like contribute?

File: packages/@aws-cdk/aws-cognito/README.md [hunk 4]
```diff
@@ -364,6 +364,20 @@ new cognito.UserPool(this, 'myuserpool', {
 
 ```
 
+Sending emails through SES verified domain, you could use the prop `sesVerifiedDomain` and pass your verified domain value.
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/README.md:367]: This comment should better explain the use case for this feature, and should mention the use case from the linked issue.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts [hunk 5]
```diff
@@ -22,14 +22,14 @@ export interface UserPoolSESOptions {
    * An optional name that should be used as the sender's name
    * along with the email.
    *
-   * @default - no name
+   * @default no name
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts:25]: There's been some confusion around the use of these `-`s: the convention we follow is to use the `-` if the text that follows the `@default` is a description of the value (eg `no name`), and to not use the `-` if the string that follows is a valid value (eg `1`, `true`). Same comment everywhere else.

**Review by comcalvi — CHANGES_REQUESTED** [review]
Looks pretty good overall. This is just a preliminary pass until the code build passes

**Comment by hassanazharkhan** [comment]
@comcalvi Would appreciate another pass

File: packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts [hunk 6]
```diff
@@ -172,7 +179,7 @@ class SESEmail extends UserPoolEmail {
       sourceArn: Stack.of(scope).formatArn({
         service: 'ses',
         resource: 'identity',
-        resourceName: encodeAndTest(this.options.fromEmail),
+        resourceName: encodeAndTest(this.options.sesVerifiedDomain ?? this.options.fromEmail),
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts:182]: If the email address has a different domain than the `sesVerifiedDomain`, a deploy time error occurs:

```
UPDATE_FAILED  | AWS::Cognito::UserPool | myuserpool01998219
Unable to send email message, please try again (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: dad6a465-2345-4d13-b1d1-45a038ba1355; Proxy: null
)
```

we need to catch this here, and throw an error if the domain doesn't match (need a unit test for this as well).

**Review by comcalvi — CHANGES_REQUESTED** [review]
Looks really good! A few small comments.

**Review by comcalvi — CHANGES_REQUESTED** [review]
Please update the PR description to describe these changes.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts (hunk 6 above)
Comment (hassanazharkhan) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-email.ts:189]: Done!

**Review by comcalvi — DISMISSED** [review]

**Comment by hassanazharkhan** [comment]
@comcalvi Can you please re-approve, not sure why the build failed so I just rebase my branch from master it is passed now!

**Review by comcalvi — APPROVED** [review]

**Pull request merged**